File: app/src/main/java/tech/pixelw/castrender/ui/PlayerActivity.java
package tech.pixelw.castrender.ui;

import android.app.Service;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.content.res.Configuration;
import android.os.Bundle;
import android.os.IBinder;
import android.text.TextUtils;
import android.view.KeyEvent;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.BaseInputConnection;
import android.widget.ImageButton;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.core.view.WindowCompat;
import androidx.core.view.WindowInsetsCompat;
import androidx.core.view.WindowInsetsControllerCompat;
import androidx.databinding.DataBindingUtil;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.exoplayer2.C;
import com.google.android.exoplayer2.MediaItem;
import com.google.android.exoplayer2.SimpleExoPlayer;
import com.google.android.exoplayer2.audio.AudioAttributes;
import com.google.android.material.dialog.MaterialAlertDialogBuilder;

import org.fourthline.cling.support.avtransport.lastchange.AVTransportVariable;
import org.fourthline.cling.support.model.Channel;
import org.fourthline.cling.support.model.TransportState;
import org.fourthline.cling.support.renderingcontrol.lastchange.ChannelVolume;
import org.fourthline.cling.support.renderingcontrol.lastchange.RenderingControlVariable;

import tech.pixelw.castrender.R;
import tech.pixelw.castrender.databinding.ActivityPlayer2Binding;
import tech.pixelw.castrender.ui.mediainfo.MediaInfoListAdapter;
import tech.pixelw.castrender.ui.mediainfo.MediaInfoRetriever;
import tech.pixelw.castrender.utils.LogUtil;
import tech.pixelw.castrender.utils.SafeZoneHelper;
import tech.pixelw.dmr_core.DLNARendererService;

public class PlayerActivity extends AppCompatActivity implements ExoRenderControlImpl.ActivityCallback {

    public static final String EXTRA_KEY_URL = "EXTRA_KEY_MEDIA_URL";
    private ActivityPlayer2Binding binding;
    private SimpleExoPlayer exoPlayer;
    private DLNARendererService.RendererServiceBinder binder;
    private static final String TAG = "PlayerActivity";

    private final ServiceConnection connection = new ServiceConnection() {
        @Override
        public void onServiceConnected(ComponentName name, IBinder service) {
            binder = (DLNARendererService.RendererServiceBinder) service;
            binder.registerController(new ExoRenderControlImpl(exoPlayer, PlayerActivity.this));
        }

        @Override
        public void onServiceDisconnected(ComponentName name) {
            binder = null;
        }
    };
    private WindowInsetsControllerCompat controllerCompat;
    private ViewGroup safeZone;
    private MediaInfoRetriever retriever;

    public static void newPlayerInstance(Context context, String url) {
        Intent intent = new Intent(context, PlayerActivity.class);
        intent.putExtra(EXTRA_KEY_URL, url);
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        context.startActivity(intent);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = DataBindingUtil.setContentView(this, R.layout.activity_player2);
        controllerCompat = new WindowInsetsControllerCompat(getWindow(), binding.exoPlayerView);
        Toolbar toolbar = binding.exoPlayerView.findViewById(R.id.player_toolbar);
        setSupportActionBar(toolbar);
        safeZone = binding.exoPlayerView.findViewById(R.id.cl_controller_safe_zone);
        if (safeZone != null) {
            SafeZoneHelper.observe(safeZone, true, true, insets -> {
                LogUtil.i(TAG, "SafeZone in pixels " + insets);
                ViewGroup.MarginLayoutParams lp = (ViewGroup.MarginLayoutParams) safeZone.getLayoutParams();
                int horizontal = Math.max(Math.max(insets.left, insets.right), Math.max(lp.leftMargin, lp.rightMargin));
                int top = Math.max(lp.topMargin, insets.top);
                int bottom = Math.max(lp.bottomMargin, insets.bottom);
                lp.setMargins(horizontal, top, horizontal, bottom);
            });
        }
        exoPlayer = new SimpleExoPlayer.Builder(this).build();
        exoPlayer.setAudioAttributes(
                new AudioAttributes.Builder().setUsage(C.USAGE_MEDIA)
                        .setContentType(C.CONTENT_TYPE_MOVIE).build(),
                true);
        binding.exoPlayerView.setPlayer(exoPlayer);
        binding.setHandler(new Handler());
        retriever = new MediaInfoRetriever(exoPlayer);
        connectToService();
        onNewIntent(getIntent());
    }

    private void connectToService() {
        Intent intent = new Intent(this, DLNARendererService.class);
        intent.putExtra("foreground", true);
        bindService(intent, connection, Service.BIND_AUTO_CREATE);
    }

    @Override
    protected void onNewIntent(Intent intent) {
        if (intent == null) return;
        super.onNewIntent(intent);
        String url = intent.getStringExtra(EXTRA_KEY_URL);
        if (!TextUtils.isEmpty(url)) {
            prepareAndPlayMedia(url);
        }
    }

    private void prepareAndPlayMedia(String url) {
        runOnUiThread(() -> {
            MediaItem mediaItem = MediaItem.fromUri(url);
            exoPlayer.setMediaItem(mediaItem);
            exoPlayer.prepare();
            exoPlayer.play();
        });
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_player, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        if (item.getItemId() == R.id.menu_media_info) {
            View view = getLayoutInflater().inflate(R.layout.layout_media_info, null);
            RecyclerView recyclerView = view.findViewById(R.id.list_media_info);
            if (recyclerView != null) {
                recyclerView.setLayoutManager(new LinearLayoutManager(this));
                // TODO: 2022/2/10 list
                MediaInfoListAdapter adapter = new MediaInfoListAdapter();
                adapter.setTrackList(retriever.mediaInfo.getTracks());
                recyclerView.setAdapter(adapter);
                AlertDialog dialog = new MaterialAlertDialogBuilder(this)
                        .setTitle("MediaInfo")
                        .setView(view)
                        .create();
                dialog.show();
            }

        } else if (item.getItemId() == R.id.menu_speed_settings) {

        }
        return true;
    }

    @Override
    protected void onStart() {
        super.onStart();
        WindowCompat.setDecorFitsSystemWindows(getWindow(), false);
        controllerCompat.hide(WindowInsetsCompat.Type.systemBars());
        controllerCompat.setSystemBarsBehavior(WindowInsetsControllerCompat.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE);
    }

    @Override
    protected void onPause() {
        super.onPause();
        exoPlayer.pause();
    }

    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {
        boolean handled = super.onKeyDown(keyCode, event);
        switch (keyCode) {
            case KeyEvent.KEYCODE_ENTER:
            case KeyEvent.KEYCODE_DPAD_CENTER:
            case KeyEvent.KEYCODE_NUMPAD_ENTER:
            case KeyEvent.KEYCODE_MEDIA_PLAY_PAUSE:
                if (exoPlayer == null) return handled;
                if (exoPlayer.isPlaying()) {
                    exoPlayer.pause();
                    notifyTransportStateChanged(TransportState.PAUSED_PLAYBACK);
                } else {
                    exoPlayer.play();
                    notifyTransportStateChanged(TransportState.PLAYING);
                }
                break;
            default:
                break;
        }
        return handled;
    }

    private void notifyTransportStateChanged(TransportState state) {
        if (binder != null) {
            binder.avTransportLastChange()
                    .setEventedValue(binder.getInstanceId(),
                            new AVTransportVariable.TransportState(state));
        }
    }

    private void notifyRenderVolumeChanged(int volume) {
        if (binder != null) {
            binder.audioControlLastChange().
                    setEventedValue(binder.getInstanceId(),
                            new RenderingControlVariable.Volume(
                                    new ChannelVolume(Channel.Master, volume)));
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        exoPlayer.stop();
        notifyTransportStateChanged(TransportState.STOPPED);
        exoPlayer.release();
        binder.unregisterController();
        unbindService(connection);
    }

    @Override
    public void finishPlayer() {
        finish();
    }

    public final class Handler {
        public void navigationClick(View view) {
            finish();
        }
    }

}

File: app/src/main/java/tech/pixelw/castrender/feature/render/PlayerActivity.kt
package tech.pixelw.castrender.feature.render

import android.content.Context
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.text.TextUtils
import android.view.KeyEvent
import android.view.Menu
import android.view.MenuItem
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.WindowCompat
import androidx.core.view.WindowInsetsCompat
import androidx.core.view.WindowInsetsControllerCompat
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.exoplayer2.ExoPlayer
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import tech.pixelw.castrender.K
import tech.pixelw.castrender.R
import tech.pixelw.castrender.databinding.ActivityPlayerBinding
import tech.pixelw.castrender.feature.mediainfo.MediaInfoListAdapter
import tech.pixelw.castrender.feature.mediainfo.MediaInfoRetriever
import tech.pixelw.castrender.feature.render.player.ExoPlayerImplementation
import tech.pixelw.castrender.feature.render.player.IPlayer
import tech.pixelw.castrender.feature.render.player.MediaPlayerImplementation
import tech.pixelw.castrender.feature.render.player.PlayerViewHelper
import tech.pixelw.castrender.feature.settings.Pref
import tech.pixelw.castrender.utils.LogUtil
import tech.pixelw.cling_common.entity.MediaEntity
import java.lang.Float.min
import kotlin.math.roundToLong

class PlayerActivity : AppCompatActivity() {

    private lateinit var binding: ActivityPlayerBinding
    private lateinit var player: IPlayer<*>
    private lateinit var keyHandler: KeyHandler
    private lateinit var osdHelper: OSDHelper
    private lateinit var controllerCompat: WindowInsetsControllerCompat
    private val viewModel by viewModels<PlayerViewModel>()
    private var retriever: MediaInfoRetriever? = null
    private var isUserDraggingSeekBar = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = DataBindingUtil.setContentView(this, R.layout.activity_player)
        binding.lifecycleOwner = this
        binding.vm = viewModel
        binding.handler = Handler()
        controllerCompat = WindowCompat.getInsetsController(window, binding.root)

        val pair = PlayerViewHelper.initPlayerView(this, binding.frPlayContainer)
        when (pair.first) {
            Pref.V_PLAYER_EXO -> {
                player = ExoPlayerImplementation(this)
                viewModel.playerType.value = getString(R.string.exo_player)
            }
            Pref.V_PLAYER_SYS -> {
                player = MediaPlayerImplementation(lifecycleScope)
                viewModel.playerType.value = getString(R.string.sys_player)
            }
        }
        player.bindView(binding.frPlayContainer.findViewById(R.id.view_player))
        if (player is ExoPlayerImplementation) {
            retriever = MediaInfoRetriever(player.playerInstance() as ExoPlayer)
        }
        viewModel.playPosition.observe(this) {
            if (isUserDraggingSeekBar) return@observe
            val duration = player.duration
            if (duration > 0) {
                val percent = player.position.toFloat() / duration.toFloat()
                viewModel.progressBarPercent.value = min(1f, percent)
            } else {
                viewModel.progressBarPercent.value = 0f
            }
        }
        keyHandler = KeyHandler(player)
        osdHelper = OSDHelper(binding.clSafezone, binding.controls)
        keyHandler.attachOsd(osdHelper)
        viewModel.connectToService(player) {
            finish()
        }
        onNewIntent(intent)
    }

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        val url = intent.getStringExtra(K.EXTRA_KEY_URL)
        if (!TextUtils.isEmpty(url)) {
            runOnUiThread {
                if (url != null) {
                    player.prepareMedia(url, playWhenReady = true)
                }
            }
        }
        val meta: MediaEntity? = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            intent.getParcelableExtra(K.EXTRA_KEY_META, MediaEntity::class.java)
        } else {
            intent.getParcelableExtra(K.EXTRA_KEY_META)
        }
        if (meta != null) {
            viewModel.media.value = meta
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_player, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == R.id.menu_media_info) {
            val view = layoutInflater.inflate(R.layout.layout_media_info, null)
            view.findViewById<RecyclerView>(R.id.list_media_info)?.apply {
                layoutManager = LinearLayoutManager(this@PlayerActivity)
                val mediaInfoListAdapter = MediaInfoListAdapter()
                mediaInfoListAdapter.setTrackList(retriever!!.mediaInfo.tracks)
                adapter = mediaInfoListAdapter
                val dialog = MaterialAlertDialogBuilder(this@PlayerActivity)
                    .setTitle("MediaInfo")
                    .setView(view)
                    .create()
                dialog.show()
            }
            // TODO: 2022/2/10 list

        }
        return true
    }

    override fun onWindowFocusChanged(hasFocus: Boolean) {
        LogUtil.d(TAG, "windowFocus->$hasFocus")
        if (hasFocus) {
            WindowCompat.setDecorFitsSystemWindows(window, false)
            controllerCompat.hide(WindowInsetsCompat.Type.systemBars())
            controllerCompat.systemBarsBehavior = WindowInsetsControllerCompat.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE
        }
    }

    override fun onPause() {
        super.onPause()
        player.pause()
    }

    override fun onKeyUp(keyCode: Int, event: KeyEvent): Boolean {
        return if (keyHandler.onKeyUp(keyCode, event)) true else super.onKeyUp(keyCode, event)
    }

    override fun onKeyDown(keyCode: Int, event: KeyEvent): Boolean {
        return if (keyHandler.onKeyDown(keyCode, event)) true else super.onKeyDown(keyCode, event)
    }

    override fun onKeyLongPress(keyCode: Int, event: KeyEvent): Boolean {
        return if (keyHandler.onKeyLongPress(keyCode, event)) true
        else super.onKeyLongPress(keyCode, event)
    }

    override fun onDestroy() {
        super.onDestroy()
        player.close()
        viewModel.disconnect()
    }

    companion object {
        private const val TAG = "PlayerActivity"

        fun newPlayerInstance(context: Context?, url: String?, entity: MediaEntity?) {
            val intent = Intent(context, PlayerActivity::class.java)
            intent.putExtra(K.EXTRA_KEY_URL, url)
            intent.putExtra(K.EXTRA_KEY_META, entity)
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            context?.startActivity(intent)
        }
    }

    inner class Handler {
        fun onSliderDragged(v: Float) {
            if (v < 0.0f) {
                isUserDraggingSeekBar = true
            } else {
                isUserDraggingSeekBar = false
                // TODO: SLY 2023/1/18
                player.seekTo((player.duration * v).roundToLong())
            }

        }
    }
}